### PR TITLE
fix: re-raise non-cross-device PermissionError in _atomic_save to prevent silent checkpoint failures

### LINKED ISSUE #21800: Checkpoints silently fail to save due to swallowed PermissionError in `_atomic_save`
### Bug description

`_atomic_save` was silently swallowing non-cross-device PermissionErrors, causing saves to report success while failing to write any checkpoints. Because training continues as if the process were successful, this is a data-loss risk.

### What version are you seeing the problem on?

v2.6

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
import os
from pathlib import Path
import sys
import tempfile
import torch
from lightning.fabric.utilities.cloud_io import _atomic_save


def main():
    with tempfile.TemporaryDirectory() as tmpdir:
        # Create a read-only directory (mode 0o555: read & execute only, no write permission)
        readonly_dir = Path(tmpdir) / "readonly"
        readonly_dir.mkdir()
        os.chmod(readonly_dir, 0o555)

        filepath = readonly_dir / "test.ckpt"
        print(f"Attempting to save checkpoint to read-only path: {filepath}")

        try:
            # _atomic_save attempts to save inside the read-only directory.
            # A PermissionError will occur inside fsspec, but without the fix,
            # _atomic_save silently swallows it and returns None.
            _atomic_save({"key": torch.tensor([1, 2, 3])}, filepath)

            print("\n❌ BUG REPRODUCED: `_atomic_save` returned silently without raising an error!")
            print(f"   Did the file actually get saved? {filepath.exists()}")
            # Exit with code 1 to indicate the bug is present
            sys.exit(1)

        except PermissionError as e:
            print(f"\n✅ EXPECTED BEHAVIOR: PermissionError was properly raised:\n   {e}")
            sys.exit(0)
        finally:
            # Restore permissions so temporary directory cleanup succeeds
            os.chmod(readonly_dir, 0o777)


if __name__ == "__main__":
    main()
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.6.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

cc @ethanwharris @lantiga

### PR DESCRIPTION
**Pull request by xiaoyaoqilan**
## Problem

`_atomic_save` catches `PermissionError` but only handles the cross-device case (`errno.EXDEV`). All other `PermissionError`s are silently swallowed — the function returns `None` as if the save succeeded, but the checkpoint was never written. Training continues without interruption, and the user only discovers the data loss later.

## Repro

```python
# Create a read-only directory
os.chmod(dir, 0o555)
_atomic_save(checkpoint, dir / "checkpoint.ckpt")
# BUG: Returns silently. No error raised. No checkpoint saved.
```

## Fix

Added `raise` after the cross-device handler so non-cross-device `PermissionError`s propagate to the caller.

**Before:** PermissionError caught → silently returns → checkpoint lost
**After:** PermissionError caught → cross-device gets upgrade message, all others re-raised → caller knows save failed

## Impact

Prevents silent data loss when checkpoints can't be written due to permissions, filesystem issues, or disk full conditions.

Fixes #21800

### CODE REVIEW

**Comment by bhimrazy** [comment]
Closing this PR as the issue is alreadybeing resolved by PR #21799

Please feel free to look into other issues, and thanks for the contribution! 🙌
